src/health_reporter.cc

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+#include "health_reporter.h"
+#include <fstream>
+#include <sstream>
+
+namespace google {
+  const std::string external_report_filename = "metadata_agent_unhealthy";
+
+  HealthReporter::HealthReporter() : HealthReporter::HealthReporter("") { }
+
+  HealthReporter::HealthReporter(const std::string &prefix) : state_prefix(prefix) {
+    health_states.push_back("watch_pods");
+    health_states.push_back("node_callback");
+  }
+
+  void HealthReporter::SetUnhealthy(const std::string &state_name) {
+    std::ofstream healthfile (Prefix(state_name));
+    healthfile << "\n";
+    healthfile.close();
+
+    ReportHealth();
+  }
+
+  bool HealthReporter::ReportHealth() {
+    if(!GetTotalHealthState()) {
+      std::ofstream healthfile (Prefix(external_report_filename));
```

Comment:
Abstracting out `touch` could be helpful to share this logic across `SetUnhealthy` and `ReportHealth`.

Reply:
I may just be confused, but is the idea to report multiple files? One for bad state as well as one for good state? Shouldn't we otherwise be doing the touch in only one place?

Reply:
Andrew and I spoke about this, and we felt that it made sense to have multiple signals that contribute to overall health. So each file is being touched, so that we can create a holistic view of what might be broken. While this isn't being fully leveraged at the moment, ANY file being touched, creates a "global" indicator of the process being healthy when its ran through `GetTotalHealthState`, does that clear things up at all?

Reply:
Synced offline. This file helps aggregate the state across multiple files to make it easy for the liveness probe. +1 to still putting the touch in a separate private method.